src/js/components/messaging/subscribe.js

```diff
@@ -0,0 +1,52 @@
+import { API } from "../../api";
+
+export const subscribe = ($customCheckbox, $checkbox) => {
+  Notification.requestPermission()
+    .then((response) => {
+      const api = new API();
+
+      if (response === "granted") {
+        navigator.serviceWorker.ready
+          .then((serviceWorkerRegistration) => {
+            api.getVAPIDKey().done((key) => {
+              const options = {
+                userVisibleOnly: true,
+                applicationServerKey: key.vapid_public_key,
+              };
+
+              serviceWorkerRegistration.pushManager
+                .subscribe(options)
+                .then((pushSubscription) => {
+                  const subscription = pushSubscription.toJSON();
+
+                  api
+                    .createPushSubscription(
+                      JSON.stringify({
+                        subscription_object: subscription,
+                      })
+                    )
+                    .then(() => {
```

Comment:
4 thens deep!

Looks like we're entering callback hell - any thoughts on how we should restructure things to escape? Something we can tackle early next year, but I'm keen to know if you have dealt with this before and want to propose a better structure

Reply:
:) Indeed. We should look at using `async/await` or alternative, refactor the code so we can for example fire of a couple and do a `Promise.all`. The latter is not possible if the sequence of events matter or depend on each other though.

I was going to do some `async/await` but seems Parcel does not play well with it just yet. There is a Babel config change and this is currently problematic. That, or I just needs to dig in a little deeper.

Reply:
https://github.com/OpenUpSA/muni-portal-frontend/issues/71

Reply:
awesome. Much appreciated